### PR TITLE
Ковальчук Александр. Задача 1. Вариант 13.  Максимальное значение элементов матрицы.

### PR DESCRIPTION
**Pull request by KADCHDR**
Решение последовательной задачи:
Проходим по каждой строке матрицы, сравнивая максимальное значение с общим максимальным значением матрицы
Решение MPI задачи:
Матрица разбивается между процессами, каждый процесс находит своё максимальное значение в выделенной ему части матрицы, затем определяется общий глобальный максимум.

### CODE REVIEW

**Review by keshaproger — APPROVED** [review]
Вроде нормально все

File: tasks/mpi/kovalchuk_a_max_of_vector_elements/func_tests/main.cpp [hunk 1]
```diff
@@ -0,0 +1,185 @@
+// Copyright 2023 Nesterov Alexander
```
Comment (keshaproger) [inline tasks/mpi/kovalchuk_a_max_of_vector_elements/func_tests/main.cpp:1]: change: // Copyright 2023 Nesterov Alexandr
to: // Copyright 2024 Kovalchuk Alexandr

**Review by keshaproger — CHANGES_REQUESTED** [review]

File: tasks/mpi/kovalchuk_a_max_of_vector_elements/include/ops_mpi.hpp [hunk 2]
```diff
@@ -0,0 +1,50 @@
+// Copyright 2023 Nesterov Alexander
```
Comment (keshaproger) [inline tasks/mpi/kovalchuk_a_max_of_vector_elements/include/ops_mpi.hpp:1]: change: // Copyright 2023 Nesterov Alexandr
to: // Copyright 2024 Kovalchuk Alexandr

File: tasks/mpi/kovalchuk_a_max_of_vector_elements/perf_tests/main.cpp [hunk 3]
```diff
@@ -0,0 +1,95 @@
+// Copyright 2023 Nesterov Alexander
```
Comment (keshaproger) [inline tasks/mpi/kovalchuk_a_max_of_vector_elements/perf_tests/main.cpp:1]: change: // Copyright 2023 Nesterov Alexandr
to: // Copyright 2024 Kovalchuk Alexandr

**Review by keshaproger — CHANGES_REQUESTED** [review]

File: tasks/mpi/kovalchuk_a_max_of_vector_elements/src/ops_mpi.cpp [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    unsigned int columns = taskData->inputs_count[1];
+    input_ = std::vector<int>(rows * columns);
+    for (unsigned int i = 0; i < rows; i++) {
+      auto* tmp_ptr = reinterpret_cast<int*>(taskData->inputs[i]);
+      for (unsigned int j = 0; j < columns; j++) {
+        input_[i * columns + j] = tmp_ptr[j];
+      }
+    }
+    for (int proc = 1; proc < world.size(); proc++) {
+      world.send(proc, 0, input_.data() + delta * proc, delta);
```
Comment (keshaproger) [inline tasks/mpi/kovalchuk_a_max_of_vector_elements/src/ops_mpi.cpp:83]: Add buffer check and use:
std::span

**Review by keshaproger — CHANGES_REQUESTED** [review]

**Review by antihype320 — APPROVED** [review]

**Comment by allnes** [comment]
@antihype320 your review is not accepted

**Comment by allnes** [comment]
@keshaproger your review is not accepted too

File: tasks/mpi/korablev_v_rect_int_mpi/func_tests/main.cpp [hunk 5]
```diff
@@ -258,4 +258,4 @@ TEST(korablev_v_rect_int, test_remainder_case) {
 
     ASSERT_NEAR(reference_result[0], global_result[0], 1e-3);
   }
-}
```
Comment (allnes) [inline tasks/mpi/korablev_v_rect_int_mpi/func_tests/main.cpp:261]: please don't change not own files

File: tasks/mpi/kovalchuk_a_max_of_vector_elements/src/ops_mpi.cpp [hunk 6]
```diff
@@ -0,0 +1,123 @@
+// Copyright 2023 Nesterov Alexander
+#include "mpi/kovalchuk_a_max_of_vector_elements/include/ops_mpi.hpp"
+
+#include <algorithm>
+#include <functional>
+#include <random>
+#include <string>
+#include <vector>
+
+std::vector<int> kovalchuk_a_max_of_vector_elements::getRandomVector(int sz, int min, int max) {
```
Comment (allnes) [inline tasks/mpi/kovalchuk_a_max_of_vector_elements/src/ops_mpi.cpp:10]: please move this functions to tests files

**Review by allnes — CHANGES_REQUESTED** [review]

File: tasks/mpi/korablev_v_rect_int_mpi/func_tests/main.cpp [hunk 7]
```diff
@@ -258,4 +258,4 @@ TEST(korablev_v_rect_int, test_remainder_case) {
...
+}
```
Comment (keshaproger) [inline tasks/mpi/korablev_v_rect_int_mpi/func_tests/main.cpp:261]: Add one more empty line for fix the bug.

**Review by keshaproger — CHANGES_REQUESTED** [review]

File: tasks/mpi/kovalchuk_a_max_of_vector_elements/func_tests/main.cpp [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+std::vector<std::vector<int>> kovalchuk_a_max_of_vector_elements::getRandomMatrix(int rows, int columns, int min,
+                                                                                  int max) {
+  std::vector<std::vector<int>> vec(rows);
+  for (int i = 0; i < rows; i++) {
+    vec[i] = kovalchuk_a_max_of_vector_elements::getRandomVector(columns, min, max);
+  }
+  return vec;
+}
+
+TEST(kovalchuk_a_max_of_vector_elements, Test_Max_10_10) {
```
Comment (myffii) [inline tasks/mpi/kovalchuk_a_max_of_vector_elements/func_tests/main.cpp:30]: add empty matrix test

File: tasks/seq/kovalchuk_a_max_of_vector_elements/func_tests/main.cpp [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+std::vector<std::vector<int>> kovalchuk_a_max_of_vector_elements_seq::getRandomMatrix(int rows, int columns, int min,
+                                                                                      int max) {
+  std::vector<std::vector<int>> vec(rows);
+  for (int i = 0; i < rows; i++) {
+    vec[i] = kovalchuk_a_max_of_vector_elements_seq::getRandomVector(columns, min, max);
+  }
+  return vec;
+}
+
+TEST(kovalchuk_a_max_of_vector_elements_seq, Test_Max_10_10) {
```
Comment (myffii) [inline tasks/seq/kovalchuk_a_max_of_vector_elements/func_tests/main.cpp:28]: add empty matrix test too

**Review by myffii — CHANGES_REQUESTED** [review]

File: tasks/mpi/kovalchuk_a_max_of_vector_elements/src/ops_mpi.cpp [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+#include <vector>
+
+bool kovalchuk_a_max_of_vector_elements::TestMPITaskSequential::pre_processing() {
+  internal_order_test();
+  // Init vectors
+  input_ = std::vector<std::vector<int>>(taskData->inputs_count[0], std::vector<int>(taskData->inputs_count[1]));
+  for (unsigned int i = 0; i < taskData->inputs_count[0]; i++) {
+    auto* tmp_ptr = reinterpret_cast<int*>(taskData->inputs[i]);
+    std::copy(tmp_ptr, tmp_ptr + taskData->inputs_count[1], input_[i].begin());
+  }
```
Comment (myffii) [inline tasks/mpi/kovalchuk_a_max_of_vector_elements/src/ops_mpi.cpp:17]: add empty matrix check

File: tasks/mpi/kovalchuk_a_max_of_vector_elements/src/ops_mpi.cpp [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+      throw std::runtime_error("Delta exceeds input vector size");
+    }
+    for (int proc = 1; proc < world.size(); proc++) {
+      std::span<int> buffer(input_.data() + delta * proc, delta);
+      world.send(proc, 0, buffer.data(), buffer.size());
+    }
+  }
+  local_input_ = std::vector<int>(delta);
+  if (world.rank() == 0) {
+    local_input_ = std::vector<int>(input_.begin(), input_.begin() + delta);
```
Comment (myffii) [inline tasks/mpi/kovalchuk_a_max_of_vector_elements/src/ops_mpi.cpp:73]: add empty matrix check

**Review by myffii — CHANGES_REQUESTED** [review]

File: tasks/mpi/kovalchuk_a_max_of_vector_elements/include/ops_mpi.hpp [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+#include "core/task/include/task.hpp"
+
+namespace kovalchuk_a_max_of_vector_elements {
+
+const int MINIMALGEN = -999;
+const int MAXIMUMGEN = 999;
+
+std::vector<int> getRandomVector(int sz, int min = MINIMALGEN, int max = MAXIMUMGEN);
+std::vector<std::vector<int>> getRandomMatrix(int rows, int columns, int min = MINIMALGEN, int max = MAXIMUMGEN);
```
Comment (aobolensk) [inline tasks/mpi/kovalchuk_a_max_of_vector_elements/include/ops_mpi.hpp:21]: Please, move to test modules

File: tasks/mpi/kovalchuk_a_max_of_vector_elements/perf_tests/main.cpp [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+  std::vector<std::vector<int>> global_matrix;
+  std::vector<int32_t> global_max(1, INT_MIN);
+  int ref = INT_MAX;
+  // Create TaskData
+  std::shared_ptr<ppc::core::TaskData> taskDataPar = std::make_shared<ppc::core::TaskData>();
+  if (world.rank() == 0) {
+    std::random_device dev;
+    std::mt19937 gen(dev());
+    int count_rows = 999;
+    int count_columns = 999;
```
Comment (aobolensk) [inline tasks/mpi/kovalchuk_a_max_of_vector_elements/perf_tests/main.cpp:41]: Too small sizes for perf test

File: tasks/mpi/kovalchuk_a_max_of_vector_elements/func_tests/main.cpp [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+    kovalchuk_a_max_of_vector_elements::TestMPITaskSequential testMpiTaskSequential(taskDataSeq);
+    ASSERT_EQ(testMpiTaskSequential.validation(), true);
+    testMpiTaskSequential.pre_processing();
+    testMpiTaskSequential.run();
+    testMpiTaskSequential.post_processing();
+    ASSERT_EQ(reference_max[0], global_max[0]);
+  }
+}
+
+TEST(kovalchuk_a_max_of_vector_elements, Test_Max_100_100) {
```
Comment (aobolensk) [inline tasks/mpi/kovalchuk_a_max_of_vector_elements/func_tests/main.cpp:118]: Too few tests, actually, please test corner cases

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: tasks/mpi/kovalchuk_a_max_of_vector_elements/include/ops_mpi.hpp (hunk 12 above)
Comment (aobolensk) [reply tasks/mpi/kovalchuk_a_max_of_vector_elements/include/ops_mpi.hpp]: This part was not fixed

**Review by aobolensk — APPROVED** [review]

**Review by keshaproger — APPROVED** [review]

**Review by allnes — APPROVED** [review]
